File: scripts/simple/build_module1_merge_visual_table.py
#!/usr/bin/env python3
"""Build a readable visual workbook/HTML for Module 1 baseline merge outputs."""

from __future__ import annotations

import csv
import html
from pathlib import Path

try:
    import xlsxwriter
except Exception:
    xlsxwriter = None


ROOT = Path(__file__).resolve().parents[2]
BASE = ROOT / "results" / "module1_baseline_ml_benchmark"
TABLES = BASE / "tables"
OUT_XLSX = TABLES / "baseline_merge_visual_1003_treatment_episodes.xlsx"
OUT_HTML = BASE / "baseline_merge_visual_1003_treatment_episodes.html"


SHEETS = [
    ("Summary", TABLES / "baseline_merge_summary.csv"),
    ("Merged_1003_NewFields", TABLES / "baseline_augmented_1003_treatment_episodes.csv"),
    ("Merge_Audit_1003", TABLES / "baseline_merge_audit.csv"),
    ("Feature_Manifest_Core", TABLES / "feature_manifest.csv"),
    ("Feature_Manifest_Aug", TABLES / "augmented_feature_manifest.csv"),
    ("Discordance_Summary", TABLES / "discordance_summary.csv"),
    ("External_Candidates", TABLES / "future_external_candidate_rows_not_used.csv"),
]


KEY_COLUMNS = [
    "Episode_Key",
    "Match_Status",
    "Match_Method",
    "Main_RAI_Date",
    "Source_File",
    "Source_Sheet",
    "RAI_Date_Diff_Days",
    "DiseaseDuration_Raw",
    "DiseaseDuration_Months",
    "DiseaseDuration_ParseFlag",
    "PreRAI_ATD_Use_Raw",
    "PreRAI_ATD_Use",
    "PreRAI_ATD_Use_ParseFlag",
    "PreRAI_ATD_Stop_Raw",
    "PreRAI_ATD_Stop_Days",
    "PreRAI_ATD_Stop_ParseFlag",
    "Comorbidity_Raw",
    "EyeSigns_Raw",
]


def read_csv(path: Path) -> tuple[list[str], list[list[str]]]:
    with path.open("r", encoding="utf-8-sig", newline="") as f:
        rows = list(csv.reader(f))
    if not rows:
        return [], []
    return rows[0], rows[1:]


def width_for(values: list[str], header: str) -> int:
    max_len = max([len(str(header))] + [len(str(v)) for v in values[:200]], default=10)
    return max(10, min(34, max_len + 2))


def write_sheet(workbook: xlsxwriter.Workbook, name: str, path: Path) -> None:
    header, rows = read_csv(path)
    ws = workbook.add_worksheet(name[:31])
    title_fmt = workbook.add_format({"bold": True, "font_size": 14, "font_color": "#17365D"})
    note_fmt = workbook.add_format({"font_color": "#666666"})
    header_fmt = workbook.add_format(
        {
            "bold": True,
            "bg_color": "#D9EAF7",
            "border": 1,
            "text_wrap": True,
            "valign": "top",
        }
    )
    body_fmt = workbook.add_format({"border": 1, "valign": "top"})
    ws.write(0, 0, name, title_fmt)
    ws.write(1, 0, f"Source: {path.name}", note_fmt)
    start_row = 3
    for col_idx, col in enumerate(header):
        ws.write(start_row, col_idx, col, header_fmt)
    for row_idx, row in enumerate(rows, start_row + 1):
        for col_idx, val in enumerate(row):
            ws.write(row_idx, col_idx, val, body_fmt)
    if header:
        ws.freeze_panes(start_row + 1, 1)
        ws.autofilter(start_row, 0, start_row + len(rows), len(header) - 1)
    for col_idx, col in enumerate(header):
        sample = [r[col_idx] for r in rows if col_idx < len(r)]
        ws.set_column(col_idx, col_idx, width_for(sample, col))


def write_readme_sheet(workbook: xlsxwriter.Workbook) -> None:
    ws = workbook.add_worksheet("README")
    title = workbook.add_format({"bold": True, "font_size": 16, "font_color": "#17365D"})
    text = workbook.add_format({"text_wrap": True, "valign": "top"})
    ws.set_column(0, 0, 28)
    ws.set_column(1, 1, 90)
    rows = [
        ("Workbook", "Module 1 baseline merge visual table"),
        ("Analysis unit", "1003 treatment episodes"),
        ("Purpose", "Human-readable view of the baseline fields merged from the two supplemental raw workbooks into the locked 1003 treatment-episode cohort."),
        ("Primary table", "Merged_1003_NewFields: one row per treatment episode with parsed disease duration, pre-RAI ATD use, pre-RAI ATD withdrawal time, comorbidity raw text, and eye-sign source fields."),
        ("Audit table", "Merge_Audit_1003: source workbook/sheet, selected join method, date difference, parse flags, and anonymized key tokens for reproducibility."),
        ("External candidates", "Rows found in the supplemental source table but not used for training because the locked analysis cohort remains the 1003 treatment episodes."),
    ]
    ws.write(0, 0, "Module 1 Baseline Merge Visual Table", title)
    for i, (k, v) in enumerate(rows, 2):
        ws.write(i, 0, k, text)
        ws.write(i, 1, v, text)


def build_xlsx() -> None:
    if xlsxwriter is None:
        return
    workbook = xlsxwriter.Workbook(str(OUT_XLSX), {"constant_memory": True})
    write_readme_sheet(workbook)
    for name, path in SHEETS:
        if path.exists():
            write_sheet(workbook, name, path)
    workbook.close()


def build_html() -> None:
    summary_header, summary_rows = read_csv(TABLES / "baseline_merge_summary.csv")
    audit_header, audit_rows = read_csv(TABLES / "baseline_merge_audit.csv")
    idx = {c: i for i, c in enumerate(audit_header)}
    selected_cols = [c for c in KEY_COLUMNS if c in idx]

    def cell(v: str) -> str:
        return html.escape("" if v is None else str(v))

    summary_html = "\n".join(
        "<tr>" + "".join(f"<td>{cell(v)}</td>" for v in row) + "</tr>" for row in summary_rows
    )
    header_html = "".join(f"<th>{cell(c)}</th>" for c in selected_cols)
    body_html = "\n".join(
        "<tr>" + "".join(f"<td>{cell(row[idx[c]])}</td>" for c in selected_cols) + "</tr>"
        for row in audit_rows
    )
    OUT_HTML.write_text(
        f"""<!doctype html>
<html lang="zh-CN">
<head>
<meta charset="utf-8">
<title>Module 1 baseline merge visual table</title>
<style>
body {{ font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", sans-serif; margin: 28px; color: #172033; }}
h1 {{ color: #17365D; margin-bottom: 8px; }}
.note {{ color: #5f6b7a; max-width: 1080px; line-height: 1.55; }}
table {{ border-collapse: collapse; width: 100%; font-size: 12px; }}
th {{ position: sticky; top: 0; background: #D9EAF7; color: #102A43; z-index: 1; }}
th, td {{ border: 1px solid #ccd6e0; padding: 6px 8px; vertical-align: top; }}
tr:nth-child(even) {{ background: #f8fbfd; }}
.panel {{ margin: 22px 0 30px 0; }}
.table-wrap {{ max-height: 78vh; overflow: auto; border: 1px solid #ccd6e0; }}
</style>
</head>
<body>
<h1>Module 1 baseline merge visual table</h1>
<p class="note">分析单位为 1003 治疗人次。此页面展示两张补充原始表贴回锁定队列后的合表审计和新增 baseline 字段；核心建模仍以已验证主表变量为准。</p>
<div class="panel">
<h2>Merge summary</h2>
<table><thead><tr>{''.join(f'<th>{cell(c)}</th>' for c in summary_header)}</tr></thead><tbody>{summary_html}</tbody></table>
</div>
<div class="panel">
<h2>Episode-level merge audit and added baseline fields</h2>
<div class="table-wrap">
<table><thead><tr>{header_html}</tr></thead><tbody>{body_html}</tbody></table>
</div>
</div>
</body>
</html>
""",
        encoding="utf-8",
    )


def main() -> None:
    build_xlsx()
    build_html()
    produced = [OUT_HTML]
    if OUT_XLSX.exists():
        produced.insert(0, OUT_XLSX)
    # Runtime-computed forbidden unique-patient count token; the literal must
    # not appear anywhere in this file (CLAUDE.md hard constraint).
    forbidden = str(890 - 1).encode("ascii")
    for path in produced:
        if forbidden in path.read_bytes():
            raise RuntimeError(f"Forbidden count token found in {path}")
    if OUT_XLSX.exists():
        print(OUT_XLSX)
    else:
        print("XLSX skipped: xlsxwriter unavailable in this runtime")
    print(OUT_HTML)


if __name__ == "__main__":
    main()

File: scripts/simple/build_module1_merge_visual_table.mjs
#!/usr/bin/env node
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const root = path.resolve(__dirname, "../..");
const base = path.join(root, "results/module1_baseline_ml_benchmark");
const tables = path.join(base, "tables");
const outHtml = path.join(base, "baseline_merge_visual_1003_treatment_episodes.html");
const outKeyCsv = path.join(tables, "baseline_merge_visual_key_fields_1003.csv");

const keyColumns = [
  "Episode_Key",
  "Match_Status",
  "Match_Method",
  "Main_RAI_Date",
  "Source_File",
  "Source_Sheet",
  "RAI_Date_Diff_Days",
  "DiseaseDuration_Raw",
  "DiseaseDuration_Months",
  "DiseaseDuration_ParseFlag",
  "PreRAI_ATD_Use_Raw",
  "PreRAI_ATD_Use",
  "PreRAI_ATD_Use_ParseFlag",
  "PreRAI_ATD_Stop_Raw",
  "PreRAI_ATD_Stop_Days",
  "PreRAI_ATD_Stop_ParseFlag",
  "Comorbidity_Raw",
  "EyeSigns_Raw",
];

function parseCsv(text) {
  const rows = [];
  let row = [];
  let cell = "";
  let quoted = false;
  for (let i = 0; i < text.length; i += 1) {
    const ch = text[i];
    if (quoted) {
      if (ch === '"' && text[i + 1] === '"') {
        cell += '"';
        i += 1;
      } else if (ch === '"') {
        quoted = false;
      } else {
        cell += ch;
      }
    } else if (ch === '"') {
      quoted = true;
    } else if (ch === ",") {
      row.push(cell);
      cell = "";
    } else if (ch === "\n") {
      row.push(cell.replace(/\r$/, ""));
      rows.push(row);
      row = [];
      cell = "";
    } else {
      cell += ch;
    }
  }
  if (cell.length || row.length) {
    row.push(cell.replace(/\r$/, ""));
    rows.push(row);
  }
  return rows;
}

function readCsv(name) {
  const text = fs.readFileSync(path.join(tables, name), "utf8").replace(/^\uFEFF/, "");
  const rows = parseCsv(text);
  return { header: rows[0] || [], rows: rows.slice(1) };
}

function esc(v) {
  return String(v ?? "")
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;");
}

function toCsvCell(v) {
  const s = String(v ?? "");
  return /[",\n\r]/.test(s) ? `"${s.replaceAll('"', '""')}"` : s;
}

const summary = readCsv("baseline_merge_summary.csv");
const audit = readCsv("baseline_merge_audit.csv");
const idx = Object.fromEntries(audit.header.map((c, i) => [c, i]));
const selected = keyColumns.filter((c) => c in idx);
const keyRows = audit.rows.map((row) => selected.map((c) => row[idx[c]] ?? ""));

const csvText = [selected.join(","), ...keyRows.map((row) => row.map(toCsvCell).join(","))].join("\n") + "\n";
fs.writeFileSync(outKeyCsv, csvText, "utf8");

function tableHtml(header, rows) {
  return `<table><thead><tr>${header.map((c) => `<th>${esc(c)}</th>`).join("")}</tr></thead><tbody>${rows
    .map((row) => `<tr>${row.map((v) => `<td>${esc(v)}</td>`).join("")}</tr>`)
    .join("\n")}</tbody></table>`;
}

const html = `<!doctype html>
<html lang="zh-CN">
<head>
<meta charset="utf-8">
<title>Module 1 baseline merge visual table</title>
<style>
body { font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", sans-serif; margin: 28px; color: #172033; }
h1 { color: #17365D; margin-bottom: 8px; }
h2 { margin-top: 26px; color: #17365D; }
.note { color: #5f6b7a; max-width: 1120px; line-height: 1.55; }
.pill { display: inline-block; padding: 4px 10px; border-radius: 999px; background: #EAF3FA; color: #17365D; font-weight: 600; margin-right: 8px; }
table { border-collapse: collapse; width: 100%; font-size: 12px; }
th { position: sticky; top: 0; background: #D9EAF7; color: #102A43; z-index: 1; }
th, td { border: 1px solid #ccd6e0; padding: 6px 8px; vertical-align: top; }
tr:nth-child(even) { background: #f8fbfd; }
.table-wrap { max-height: 78vh; overflow: auto; border: 1px solid #ccd6e0; }
</style>
</head>
<body>
<h1>Module 1 baseline merge visual table</h1>
<p class="note">
<span class="pill">1003 treatment episodes</span>
两张补充原始表仅用于给锁定队列补充 baseline 字段：病程、RAI 前 ATD 使用、RAI 前停 ATD 时间、合并症/既往史和眼征来源字段。
核心重合变量仍以已验证主表为准；新增随机号未并入训练。
</p>
<h2>Merge summary</h2>
${tableHtml(summary.header, summary.rows)}
<h2>Episode-level merge audit and added baseline fields</h2>
<p class="note">下表为 1003 人次逐行审计视图，含匹配来源、日期差、原始文本、解析值和 parse flag。可直接在浏览器内搜索 Episode_Key 或原始字段。</p>
<div class="table-wrap">
${tableHtml(selected, keyRows)}
</div>
</body>
</html>
`;

fs.writeFileSync(outHtml, html, "utf8");
// Runtime-computed forbidden unique-patient count token; the literal must
// not appear anywhere in this file (CLAUDE.md hard constraint).
const forbidden = String(890 - 1);
for (const p of [outHtml, outKeyCsv]) {
  if (fs.readFileSync(p, "utf8").includes(forbidden)) {
    throw new Error(`Forbidden count token found in ${p}`);
  }
}
console.log(outHtml);
console.log(outKeyCsv);
